File: ChaCha20Poly1305StringEncryption/Chacha20Poly1305StringEncryption.cs
using System;
using System.Security.Cryptography;
using NSec.Cryptography;

public static class Program
{
	public static void Main() {
	Console.WriteLine("ChaCha20-Poly1305 String encryption with random key full");
    // you need the library https://www.nuget.org/packages/NSec.Cryptography version 20.2.0
	// documentation: https://nsec.rocks/
	// https://github.com/ektrah/nsec

    string plaintext = "The quick brown fox jumps over the lazy dog";
	Console.WriteLine("plaintext: " + plaintext);

	byte[] encryptionKey = GenerateRandomKey();
	string encryptionKeyBase64 = Base64Encoding(encryptionKey);
	Console.WriteLine("encryptionKey (Base64): " + encryptionKeyBase64);

    // encrypt

	string ciphertextBase64 = Chacha20Poly1305EncryptToBase64(encryptionKey, plaintext);
	//string ciphertextBase64 = SecretboxEncryptionToBase64(encryptionKeyBase64, plaintext);
    Console.WriteLine("\n* * * encryption * * *");
    Console.WriteLine("all data are in Base64 encoding");
    Console.WriteLine("ciphertext:  " + ciphertextBase64);
    Console.WriteLine("output is (Base64) nonce : (Base64) ciphertext");

    // decrypt
    string ciphertextReceivedBase64 = ciphertextBase64;
	string decryptionKeyBase64 = encryptionKeyBase64;
	byte[] decryptionKey = Base64Decoding(decryptionKeyBase64);
	string decryptedtext = Chacha20Poly1305DecryptFromBase64(decryptionKey, ciphertextReceivedBase64);
    Console.WriteLine("\n* * * decryption * * *");
    Console.WriteLine("ciphertext:  " + ciphertextReceivedBase64);
    Console.WriteLine("input is (Base64) nonce : (Base64) ciphertext");
    Console.WriteLine("decrypt.text:" + decryptedtext);
}

	static string Chacha20Poly1305EncryptToBase64(byte[] encryptionKey, string plaintext) {
		byte[] nonce = GenerateRandomNonce();
		//byte[] nonce = new byte[12];
		byte[] data = System.Text.Encoding.UTF8.GetBytes(plaintext);
		byte[] aad = new byte[0];
		// Create the primitive
		// select the Ed25519 signature algorithm
		var alg = AeadAlgorithm.ChaCha20Poly1305;
		using var key = Key.Import(alg, encryptionKey, KeyBlobFormat.RawSymmetricKey);
		var ciphertextWithTag = alg.Encrypt(key, new Nonce(nonce, 0), aad, data);
		// split ciphertextWitTag in ciphertext and poly1305Tag
		int ciphertextWithTagLength = ciphertextWithTag.Length;
		byte[] ciphertext = new byte[ciphertextWithTagLength - 16];
		byte[] poly1305Tag = new byte[16];
		Array.Copy(ciphertextWithTag, 0, ciphertext, 0, (ciphertextWithTagLength - 16));
		Array.Copy(ciphertextWithTag, (ciphertextWithTagLength - 16), poly1305Tag, 0, 16);
		return Base64Encoding(nonce) + ":" + Base64Encoding(ciphertext) + ":" + Base64Encoding(poly1305Tag);
	}

	static string Chacha20Poly1305DecryptFromBase64(byte[] encryptionKey, string ciphertextReceivedBase64) {
		String[] parts = ciphertextReceivedBase64.Split(':');
	  	byte[] nonce = Base64Decoding(parts[0]);
	  	byte[] ciphertext = Base64Decoding(parts[1]);
		byte[] poly1305Tag = Base64Decoding(parts[2]);
		byte[] ciphertextWithTag = new byte[(ciphertext.Length + poly1305Tag.Length)];
		Array.Copy(ciphertext, 0, ciphertextWithTag, 0, ciphertext.Length);
		Array.Copy(poly1305Tag, 0, ciphertextWithTag, ciphertext.Length, poly1305Tag.Length);
		byte[] aad = new byte[0];
		var alg = AeadAlgorithm.ChaCha20Poly1305;
		using var key = Key.Import(alg, encryptionKey, KeyBlobFormat.RawSymmetricKey);
		byte[] decrypted = new byte[ciphertext.Length];
		bool success = alg.Decrypt(key, new Nonce(nonce, 0), aad, ciphertextWithTag, decrypted);
		return System.Text.Encoding.UTF8.GetString(decrypted, 0, decrypted.Length);;
	}

	static byte[] GenerateRandomKey() {
    	RNGCryptoServiceProvider rngCsp = new RNGCryptoServiceProvider();
    	byte[] key = new byte[32];
    	rngCsp.GetBytes(key);
    	return key;
	}

    static byte[] GenerateRandomNonce() {
       RNGCryptoServiceProvider rngCsp = new RNGCryptoServiceProvider();
       byte[] nonce = new byte[12];
       rngCsp.GetBytes(nonce);
       return nonce;
    }

	private static string Base64Encoding(byte[] input) {
    	return Convert.ToBase64String(input);
  	}

  	private static byte[] Base64Decoding(String input) {
    	return Convert.FromBase64String(input);
  	}
}

File: ChaCha20Poly1305StringEncryption/Chacha20Poly1305StringEncryption.dart
import 'dart:convert';
import 'dart:math';
import 'dart:typed_data';
import "package:pointycastle/export.dart";

main() {
/* add in pubspec.yaml:
dependencies:
  pointycastle: ^3.1.1
 */

  print('ChaCha20-Poly1305 String encryption with random key full\n');
  final plaintext = 'The quick brown fox jumps over the lazy dog';
  print('plaintext: ' + plaintext);

  // generate random key
  var encryptionKey = generateRandomAesKey();
  var encryptionKeyBase64 = base64Encoding(encryptionKey);
  print('encryptionKey (Base64): ' + encryptionKeyBase64);

  print('\n* * * Encryption * * *');
  var ciphertextBase64 = chacha20Poly1305EncryptToBase64(encryptionKey, plaintext);
  print('ciphertext (Base64): ' + ciphertextBase64);
  print('output is (Base64) nonce : (Base64) ciphertext : (Base64) poly1305Tag');

  print('\n* * * Decryption * * *');
  var ciphertextDecryptionBase64 = ciphertextBase64;
  print('ciphertext (Base64): ' + ciphertextDecryptionBase64);
  print('input is (Base64) nonce : (Base64) ciphertext : (Base64) poly1305Tag');
  var decryptedtext = chacha20Poly1305DecryptFromBase64(encryptionKey, ciphertextDecryptionBase64);
  print('plaintext:  ' + decryptedtext);
}

String chacha20Poly1305EncryptToBase64(Uint8List key, String plaintext) {
  var plaintextUint8 = createUint8ListFromString(plaintext);
  final nonce = generateRandomNonce();
  var parameters = AEADParameters(KeyParameter(key), 128, nonce, Uint8List(0));
  var chaChaEngine = ChaCha20Poly1305(ChaCha7539Engine(), Poly1305())
    ..init(true, parameters);
  var enc = Uint8List(chaChaEngine.getOutputSize(plaintextUint8.length));
  var len = chaChaEngine.processBytes(plaintextUint8, 0, plaintextUint8.length, enc, 0);
  len += chaChaEngine.doFinal(enc, len);
  if (enc.length != len) {
    throw StateError('');
  }
  var ciphertextWithTag = enc;
  var ciphertextWithTagLength = ciphertextWithTag.lengthInBytes;
  var ciphertextLength = ciphertextWithTagLength - 16; // 16 bytes = 128 bit tag length
  var ciphertext = Uint8List.sublistView(ciphertextWithTag, 0, ciphertextLength);
  var macTag = Uint8List.sublistView(ciphertextWithTag, ciphertextLength, ciphertextWithTagLength);
  final nonceBase64 = base64.encode(nonce);
  final ciphertextBase64 = base64.encode(ciphertext);
  final poly1305TagBase64 = base64.encode(macTag);
  return nonceBase64 + ':' + ciphertextBase64 + ':' + poly1305TagBase64;
}

String chacha20Poly1305DecryptFromBase64(Uint8List key, String data) {
  var parts = data.split(':');
  var nonce = base64.decode(parts[0]);
  var ciphertext = base64.decode(parts[1]);
  var poly1305Tag = base64.decode(parts[2]);
  var bb = BytesBuilder();
  bb.add(ciphertext);
  bb.add(poly1305Tag);
  var ciphertextWithTag = bb.toBytes();
  var parameters = AEADParameters(KeyParameter(key), 128, nonce, Uint8List(0));
  var chaChaEngine = ChaCha20Poly1305(ChaCha7539Engine(), Poly1305())
    ..init(false, parameters);
  var dec = Uint8List(chaChaEngine.getOutputSize(ciphertextWithTag.length));
  var len = chaChaEngine.processBytes(ciphertextWithTag, 0, ciphertextWithTag.length, dec, 0);
  len += chaChaEngine.doFinal(dec, len);
  var mac = chaChaEngine.mac;
  return new String.fromCharCodes(dec);
}

Uint8List generateRandomAesKey() {
  final _sGen = Random.secure();
  final _seed =
  Uint8List.fromList(List.generate(32, (n) => _sGen.nextInt(255)));
  SecureRandom sec = SecureRandom("Fortuna")..seed(KeyParameter(_seed));
  return sec.nextBytes(32);
}

Uint8List generateRandomNonce() {
  final _sGen = Random.secure();
  final _seed =
  Uint8List.fromList(List.generate(32, (n) => _sGen.nextInt(255)));
  SecureRandom sec = SecureRandom("Fortuna")..seed(KeyParameter(_seed));
  return sec.nextBytes(12);
}

Uint8List createUint8ListFromString(String s) {
  var ret = new Uint8List(s.length);
  for (var i = 0; i < s.length; i++) {
    ret[i] = s.codeUnitAt(i);
  }
  return ret;
}

String base64Encoding(Uint8List input) {
  return base64.encode(input);
}

Uint8List base64Decoding(String input) {
  return base64.decode(input);
}
